### PR TITLE
Refactor `MetaproteomicsAnalysis` slots and implement migrator

### PR DESCRIPTION
**Pull request by kheal**
This will address https://github.com/microbiomedata/nmdc-schema/issues/2157.

Full discussion is here for interested parties: https://github.com/microbiomedata/nmdc-schema/discussions/2152.
Overall, this PR will significantly minimize the proteomics data stored as metadata by refactoring the `MetaproteomicsAnalysis` class's contents.  

This PR 
1) Removes (deprecates) the following classes and slots:
`PeptideQuantification`,  `all_proteins`, `best_protein`, `min_q_value`, `peptide_sequence`, `peptide_sum_masic_abundance`, `peptide_sequence_count`.
2) Renames`ProteinQuantification` to `ProteinIdentification` to better capture content.
3) Add one slot to the `MetaproteomicsAnalysis` class - `has_protein_identifications`.  This slot's range will be `ProteinIdentification` and that class will be populated from the `best_protein` and `peptide_sequence_count` slots from the now deprecated `PeptideQuantification` class.  
4) Adds valid and invalid examples for the `MetaproteomicsAnalysis` class (previously missing), based off of real data currently in Mongo.  
5) Constrains the structured pattern on the new `razor_protein` slot (to address https://github.com/microbiomedata/nmdc-schema/issues/2028)
6) Includes a migrator for migrating current data to match the desired structure implemented in this PR.  

This PR was moved from https://github.com/microbiomedata/berkeley-schema-fy24/pull/236.

# Reviewers
To review: mslarae13, eecavanna, SamuelPurvine 
To keep informed: naglepuff, aclum, mbthornton-lbl, corilo, sujaypatil96, brynnz22

# PR Information
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Documentation
- [x] Schema change: Structure and content
  - deleted a `class` and `slots`
  - added a `slot` to a `class`
     
## Description
See above

## Related Issues / Discussions
- Related Issue(s)/Discussions 
        - https://github.com/microbiomedata/nmdc-schema/issues/2157 (will be closed after this is merged in)
        - https://github.com/microbiomedata/nmdc-schema/issues/2028 (will be closed after this is merged in)
        - https://github.com/microbiomedata/nmdc-aggregator/issues/13 (will be unblocked after this is merged in)
        - https://github.com/microbiomedata/nmdc-schema/discussions/2152 (discussion of plan for this PR)

## Did you add/update any tests?
- [x] Yes
Added valid and invalid examples as well a doctest for migrator

## Could this schema change make it so any valid data becomes invalid?
- [x] Yes _(A migrator is required)_

## If you answered "Yes", does this PR branch include that migrator?
- [x] Yes

## Does this PR have any downstream implications?
- [x] Yes
This work will need to be followed by an update to the aggregator script, an update to the structure of the aggregation tables in the schema and associated migrator, and an update to the DataPortal's handling of the KEGG terms from the aggregations.

### CODE REVIEW

File: nmdc_schema/migrators/migrator_from_X_to_PR236.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                        protein_dict[razor_protein]["unique_peptide_count"] += 1
+                        protein_dict[razor_protein]["protein_spectral_count"] += peptide_iden.get("peptide_spectral_count")
+                    else:
+                        protein_dict[razor_protein] = {"type": "ProteinIdentification", "razor_protein": razor_protein, "unique_peptide_count": 1, "protein_spectral_count": peptide_iden.get("peptide_spectral_count")}
+                
+                # convert dict to list
+                protein_list = [x for x in protein_dict.values()]
+                
+                # order alphabetically so tests are reproducible
+                workflow_dict['has_protein_identifications'] = protein_list
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_X_to_PR236.py:62]: It's not clear to me how this assignment orders something alphabetically.

**Review by eecavanna — APPROVED** [review]
Migrator looks great! Thanks for including a variety of doctests.

**Comment by kheal** [comment]
Tested the migrator, now that https://github.com/microbiomedata/issues/issues/897 is closed, this PR is ready.

File: src/schema/core.yaml [hunk 2]
```diff
@@ -1677,27 +1674,51 @@ slots:
...
     multivalued: true
+    deprecated: "not used. 2024-10.  See https://github.com/microbiomedata/nmdc-schema/issues/2157"
+
   best_protein:
     description: the specific protein identifier most correctly associated with the peptide sequence
     range: GeneProduct
+    deprecated: "not used. 2024-10.  See https://github.com/microbiomedata/nmdc-schema/issues/2157"
+
+  razor_protein:
+    description: the specific protein identifier identified in the metaproteomics analysis that is most correctly grouped to its associated peptide sequences
```
Comment (SamuelPurvine) [inline src/schema/core.yaml:1695]: Would "the metaproteomics analysis protein identifier derived from parsimonious peptide sequence association" work as a description?

File: src/schema/core.yaml [hunk 3]
```diff
@@ -1709,16 +1730,21 @@ slots:
       - "key set to false due to rare collisions: Pletnev I, Erin A, McNaught A, Blinov K, Tchekhovskoi D, Heller S (2012) InChIKey collision resistance: an experimental testing. J Cheminform. 4:12"
   inchi:
     range: string
-  peptide_sequence_count:
+  
+  unique_peptide_count:
     description: count of peptide sequences grouped to the best_protein
```
Comment (SamuelPurvine) [inline src/schema/core.yaml:1735]: count of peptide sequences grouped to a given razor_protein

File: src/schema/core.yaml [hunk 4]
```diff
@@ -1709,16 +1730,21 @@ slots:
...
     range: integer
+
   protein_spectral_count:
     description: sum of filter passing MS2 spectra associated with the best protein within a given LC-MS/MS data file
```
Comment (SamuelPurvine) [inline src/schema/core.yaml:1739]: sum of filter passing MS2 spectra associated with a given razor protein within a given LC-MS/MS data file

**Review by SamuelPurvine — COMMENTED** [review]
Made some suggestions for verbiage, otherwise looks fine to me

File: src/schema/core.yaml (hunk 2 above)
Comment (kheal) [reply src/schema/core.yaml:1695]: I will update language to indicate that these are determined using principle of parsimony

**Comment by kheal** [comment]
@picowatt - Note that once this PR is in, the scripts that write the jsons for the MetaProteomicsAnalysis records will need to be significantly altered to match the new format.  The migrator will take care of existing records.

**Comment by kheal** [comment]
Putting this back into draft until decision is reached re: mongo data applicability.

**Comment by kheal** [comment]
See #2251

**Pull request closed without merging**